File: sdks/python/src/verinode/services/wallet.py
"""
Wallet service for managing blockchain wallets.
"""

import logging
from typing import Optional, List, Dict, Any, Callable
from ..types import (
    Wallet, WalletType, NetworkType, WalletConnectRequest
)
from ..exceptions import VerinodeWalletError, VerinodeError
from ..utils import HTTPClient, WebSocketClient


class WalletService:
    """
    Service for managing blockchain wallets in the Verinode system.
    
    Provides methods to connect, manage, and interact with various wallet providers.
    """
    
    def __init__(self, http_client: HTTPClient, ws_client: WebSocketClient, config):
        """
        Initialize wallet service.
        
        Args:
            http_client: HTTP client for API requests
            ws_client: WebSocket client for real-time updates
            config: Configuration object
        """
        self.http_client = http_client
        self.ws_client = ws_client
        self.config = config
        self.logger = logging.getLogger(__name__)
        self._connected_wallets: Dict[str, Wallet] = {}
        self._wallet_listeners: Dict[str, List[Callable]] = {}
    
    async def connect(self, request: WalletConnectRequest) -> Wallet:
        """
        Connect a wallet.
        
        Args:
            request: Wallet connection request
            
        Returns:
            Connected wallet object
            
        Raises:
            VerinodeWalletError: If wallet connection fails
        """
        try:
            response = await self.http_client.post(
                "/wallets/connect",
                json=request.dict(exclude_none=True)
            )
            
            if response.get("success"):
                wallet_data = response["data"]
                wallet = Wallet(**wallet_data)
                self._connected_wallets[wallet.id] = wallet
                self.logger.info(f"Connected wallet: {wallet.id} ({wallet.wallet_type})")
                return wallet
            else:
                raise VerinodeWalletError(
                    response.get("error", "Failed to connect wallet"),
                    wallet_type=request.wallet_type.value,
                    error_code="WALLET_CONNECT_FAILED"
                )
                
        except Exception as e:
            if isinstance(e, VerinodeError):
                raise
            raise VerinodeWalletError(
                f"Failed to connect wallet: {str(e)}",
                wallet_type=request.wallet_type.value
            )
    
    async def disconnect(self, wallet_id: str) -> bool:
        """
        Disconnect a wallet.
        
        Args:
            wallet_id: Wallet ID to disconnect
            
        Returns:
            True if disconnected successfully
            
        Raises:
            VerinodeWalletError: If wallet disconnection fails
        """
        try:
            response = await self.http_client.post(f"/wallets/{wallet_id}/disconnect")
            
            if response.get("success"):
                if wallet_id in self._connected_wallets:
                    del self._connected_wallets[wallet_id]
                self.logger.info(f"Disconnected wallet: {wallet_id}")
                return True
            else:
                raise VerinodeWalletError(
                    response.get("error", "Failed to disconnect wallet"),
                    error_code="WALLET_DISCONNECT_FAILED"
                )
                
        except Exception as e:
            if isinstance(e, VerinodeError):
                raise
            raise VerinodeWalletError(f"Failed to disconnect wallet {wallet_id}: {str(e)}")
    
    async def get(self, wallet_id: str) -> Wallet:
        """
        Get wallet information.
        
        Args:
            wallet_id: Wallet ID
            
        Returns:
            Wallet object
            
        Raises:
            VerinodeWalletError: If wallet not found
        """
        try:
            response = await self.http_client.get(f"/wallets/{wallet_id}")
            
            if response.get("success"):
                wallet_data = response["data"]
                wallet = Wallet(**wallet_data)
                return wallet
            else:
                raise VerinodeWalletError(
                    response.get("error", "Wallet not found"),
                    error_code="WALLET_NOT_FOUND"
                )
                
        except Exception as e:
            if isinstance(e, VerinodeError):
                raise
            raise VerinodeWalletError(f"Failed to get wallet {wallet_id}: {str(e)}")
    
    async def list(self, user_id: Optional[str] = None) -> List[Wallet]:
        """
        List wallets.
        
        Args:
            user_id: Optional user ID filter
            
        Returns:
            List of wallets
        """
        try:
            params = {}
            if user_id:
                params["user_id"] = user_id
            
            response = await self.http_client.get("/wallets", params=params)
            
            if response.get("success"):
                wallets = [Wallet(**w) for w in response["data"]]
                return wallets
            else:
                raise VerinodeWalletError(
                    response.get("error", "Failed to list wallets"),
                    error_code="WALLET_LIST_FAILED"
                )
                
        except Exception as e:
            if isinstance(e, VerinodeError):
                raise
            raise VerinodeWalletError(f"Failed to list wallets: {str(e)}")
    
    async def get_balance(self, wallet_id: str) -> str:
        """
        Get wallet balance.
        
        Args:
            wallet_id: Wallet ID
            
        Returns:
            Balance as string
            
        Raises:
            VerinodeWalletError: If balance retrieval fails
        """
        try:
            response = await self.http_client.get(f"/wallets/{wallet_id}/balance")
            
            if response.get("success"):
                return response["data"]["balance"]
            else:
                raise VerinodeWalletError(
                    response.get("error", "Failed to get wallet balance"),
                    error_code="WALLET_BALANCE_FAILED"
                )
                
        except Exception as e:
            if isinstance(e, VerinodeError):
                raise
            raise VerinodeWalletError(f"Failed to get balance for wallet {wallet_id}: {str(e)}")
    
    async def send_transaction(
        self,
        wallet_id: str,
        to_address: str,
        amount: str,
        memo: Optional[str] = None
    ) -> Dict[str, Any]:
        """
        Send a transaction from a wallet.
        
        Args:
            wallet_id: Wallet ID
            to_address: Recipient address
            amount: Amount to send
            memo: Optional transaction memo
            
        Returns:
            Transaction information
            
        Raises:
            VerinodeWalletError: If transaction fails
        """
        try:
            response = await self.http_client.post(
                f"/wallets/{wallet_id}/send",
                json={
                    "to_address": to_address,
                    "amount": amount,
                    "memo": memo
                }
            )
            
            if response.get("success"):
                self.logger.info(f"Transaction sent from wallet {wallet_id}")
                return response["data"]
            else:
                raise VerinodeWalletError(
                    response.get("error", "Failed to send transaction"),
                    error_code="WALLET_TRANSACTION_FAILED"
                )
                
        except Exception as e:
            if isinstance(e, VerinodeError):
                raise
            raise VerinodeWalletError(f"Failed to send transaction from wallet {wallet_id}: {str(e)}")
    
    async def sign_message(
        self,
        wallet_id: str,
        message: str
    ) -> str:
        """
        Sign a message with a wallet.
        
        Args:
            wallet_id: Wallet ID
            message: Message to sign
            
        Returns:
            Signature
            
        Raises:
            VerinodeWalletError: If signing fails
        """
        try:
            response = await self.http_client.post(
                f"/wallets/{wallet_id}/sign",
                json={"message": message}
            )
            
            if response.get("success"):
                return response["data"]["signature"]
            else:
                raise VerinodeWalletError(
                    response.get("error", "Failed to sign message"),
                    error_code="WALLET_SIGN_FAILED"
                )
                
        except Exception as e:
            if isinstance(e, VerinodeError):
                raise
            raise VerinodeWalletError(f"Failed to sign message with wallet {wallet_id}: {str(e)}")
    
    async def verify_message(
        self,
        public_key: str,
        message: str,
        signature: str
    ) -> bool:
        """
        Verify a message signature.
        
        Args:
            public_key: Public key of signer
            message: Original message
            signature: Signature to verify
            
        Returns:
            True if signature is valid
        """
        try:
            response = await self.http_client.post(
                "/wallets/verify",
                json={
                    "public_key": public_key,
                    "message": message,
                    "signature": signature
                }
            )
            
            if response.get("success"):
                return response["data"]["valid"]
            else:
                raise VerinodeWalletError(
                    response.get("error", "Failed to verify message"),
                    error_code="WALLET_VERIFY_FAILED"
                )
                
        except Exception as e:
            if isinstance(e, VerinodeError):
                raise
            raise VerinodeWalletError(f"Failed to verify message: {str(e)}")
    
    def add_listener(self, event_type: str, callback: Callable):
        """
        Add event listener for wallet events.
        
        Args:
            event_type: Type of event to listen for
            callback: Callback function
        """
        if event_type not in self._wallet_listeners:
            self._wallet_listeners[event_type] = []
        self._wallet_listeners[event_type].append(callback)
    
    def remove_listener(self, event_type: str, callback: Callable):
        """
        Remove event listener.
        
        Args:
            event_type: Type of event
            callback: Callback function to remove
        """
        if event_type in self._wallet_listeners:
            try:
                self._wallet_listeners[event_type].remove(callback)
            except ValueError:
                pass
    
    async def subscribe_to_wallet_events(self, wallet_id: str):
        """
        Subscribe to real-time wallet events.
        
        Args:
            wallet_id: Wallet ID to subscribe to
        """
        if not self.ws_client:
            raise VerinodeWalletError(
                "WebSocket client not available",
                error_code="WS_CLIENT_UNAVAILABLE"
            )
        
        filters = {"wallet_id": wallet_id}
        await self.ws_client.subscribe(filters)
    
    def get_connected_wallets(self) -> List[Wallet]:
        """
        Get list of currently connected wallets.
        
        Returns:
            List of connected wallets
        """
        return list(self._connected_wallets.values())
    
    def is_connected(self, wallet_id: str) -> bool:
        """
        Check if a wallet is connected.
        
        Args:
            wallet_id: Wallet ID
            
        Returns:
            True if connected
        """
        return wallet_id in self._connected_wallets
    
    async def switch_network(self, wallet_id: str, network: NetworkType) -> Wallet:
        """
        Switch wallet network.
        
        Args:
            wallet_id: Wallet ID
            network: New network type
            
        Returns:
            Updated wallet object
        """
        try:
            response = await self.http_client.post(
                f"/wallets/{wallet_id}/switch-network",
                json={"network": network.value}
            )
            
            if response.get("success"):
                wallet_data = response["data"]
                wallet = Wallet(**wallet_data)
                if wallet_id in self._connected_wallets:
                    self._connected_wallets[wallet_id] = wallet
                self.logger.info(f"Switched wallet {wallet_id} to {network.value}")
                return wallet
            else:
                raise VerinodeWalletError(
                    response.get("error", "Failed to switch network"),
                    error_code="WALLET_NETWORK_SWITCH_FAILED"
                )
                
        except Exception as e:
            if isinstance(e, VerinodeError):
                raise
            raise VerinodeWalletError(f"Failed to switch network for wallet {wallet_id}: {str(e)}")
    
    async def get_transaction_history(
        self,
        wallet_id: str,
        limit: int = 50,
        offset: int = 0
    ) -> List[Dict[str, Any]]:
        """
        Get transaction history for a wallet.
        
        Args:
            wallet_id: Wallet ID
            limit: Maximum number of transactions
            offset: Offset for pagination
            
        Returns:
            List of transactions
        """
        try:
            params = {"limit": limit, "offset": offset}
            response = await self.http_client.get(f"/wallets/{wallet_id}/transactions", params=params)
            
            if response.get("success"):
                return response["data"]
            else:
                raise VerinodeWalletError(
                    response.get("error", "Failed to get transaction history"),
                    error_code="WALLET_TRANSACTIONS_FAILED"
                )
                
        except Exception as e:
            if isinstance(e, VerinodeError):
                raise
            raise VerinodeWalletError(f"Failed to get transaction history for wallet {wallet_id}: {str(e)}")

File: sdks/go/pkg/verinode/services/wallet.go
package services

import (
	"context"
	"fmt"
	"log"

	"github.com/Great-2025/verinode-go/pkg/verinode/types"
)

// WalletService handles wallet-related operations
type WalletService struct {
	httpClient HTTPClient
	wsClient   WebSocketClient
	config     ConfigProvider
}

// NewWalletService creates a new wallet service
func NewWalletService(httpClient HTTPClient, wsClient WebSocketClient, config ConfigProvider) *WalletService {
	return &WalletService{
		httpClient: httpClient,
		wsClient:   wsClient,
		config:     config,
	}
}

// Connect connects a wallet
func (s *WalletService) Connect(ctx context.Context, req *types.WalletConnectRequest) (*types.Wallet, error) {
	var resp types.WalletResponse
	err := s.httpClient.Post(ctx, "/wallets/connect", req, &resp)
	if err != nil {
		return nil, fmt.Errorf("failed to connect wallet: %w", err)
	}
	
	if !resp.Success {
		return nil, fmt.Errorf("failed to connect wallet: %s", safeString(resp.Error))
	}
	
	if s.config.IsLoggingEnabled() {
		log.Printf("Connected wallet: %s (%s)", resp.Data.ID, resp.Data.WalletType)
	}
	
	return &resp.Data, nil
}

// Disconnect disconnects a wallet
func (s *WalletService) Disconnect(ctx context.Context, walletID string) error {
	var resp types.APIResponse
	err := s.httpClient.Post(ctx, fmt.Sprintf("/wallets/%s/disconnect", walletID), nil, &resp)
	if err != nil {
		return fmt.Errorf("failed to disconnect wallet %s: %w", walletID, err)
	}
	
	if !resp.Success {
		return fmt.Errorf("failed to disconnect wallet: %s", safeString(resp.Error))
	}
	
	if s.config.IsLoggingEnabled() {
		log.Printf("Disconnected wallet: %s", walletID)
	}
	
	return nil
}

// Get retrieves wallet information
func (s *WalletService) Get(ctx context.Context, walletID string) (*types.Wallet, error) {
	var resp types.WalletResponse
	err := s.httpClient.Get(ctx, fmt.Sprintf("/wallets/%s", walletID), &resp)
	if err != nil {
		return nil, fmt.Errorf("failed to get wallet %s: %w", walletID, err)
	}
	
	if !resp.Success {
		return nil, fmt.Errorf("wallet not found: %s", safeString(resp.Error))
	}
	
	return &resp.Data, nil
}

// List lists wallets with optional filtering
func (s *WalletService) List(ctx context.Context, userID *string) ([]types.Wallet, error) {
	params := make(map[string]string)
	if userID != nil {
		params["user_id"] = *userID
	}
	
	var resp types.WalletsResponse
	err := s.httpClient.Get(ctx, "/wallets", &resp, params...)
	if err != nil {
		return nil, fmt.Errorf("failed to list wallets: %w", err)
	}
	
	if !resp.Success {
		return nil, fmt.Errorf("failed to list wallets: %s", safeString(resp.Error))
	}
	
	return resp.Data, nil
}

// GetBalance retrieves wallet balance
func (s *WalletService) GetBalance(ctx context.Context, walletID string) (string, error) {
	var resp types.BalanceResponse
	err := s.httpClient.Get(ctx, fmt.Sprintf("/wallets/%s/balance", walletID), &resp)
	if err != nil {
		return "", fmt.Errorf("failed to get wallet balance %s: %w", walletID, err)
	}
	
	if !resp.Success {
		return "", fmt.Errorf("failed to get wallet balance: %s", safeString(resp.Error))
	}
	
	return resp.Data.Balance, nil
}

// SendTransaction sends a transaction from a wallet
func (s *WalletService) SendTransaction(ctx context.Context, walletID, toAddress, amount string, memo *string) (map[string]interface{}, error) {
	req := types.TransactionRequest{
		ToAddress: toAddress,
		Amount:    amount,
		Memo:      memo,
	}
	
	var resp types.TransactionResponse
	err := s.httpClient.Post(ctx, fmt.Sprintf("/wallets/%s/send", walletID), req, &resp)
	if err != nil {
		return nil, fmt.Errorf("failed to send transaction from wallet %s: %w", walletID, err)
	}
	
	if !resp.Success {
		return nil, fmt.Errorf("failed to send transaction: %s", safeString(resp.Error))
	}
	
	if s.config.IsLoggingEnabled() {
		log.Printf("Transaction sent from wallet %s", walletID)
	}
	
	return resp.Data, nil
}

// SignMessage signs a message with a wallet
func (s *WalletService) SignMessage(ctx context.Context, walletID, message string) (string, error) {
	req := types.SignMessageRequest{
		Message: message,
	}
	
	var resp types.SignMessageResponse
	err := s.httpClient.Post(ctx, fmt.Sprintf("/wallets/%s/sign", walletID), req, &resp)
	if err != nil {
		return "", fmt.Errorf("failed to sign message with wallet %s: %w", walletID, err)
	}
	
	if !resp.Success {
		return "", fmt.Errorf("failed to sign message: %s", safeString(resp.Error))
	}
	
	return resp.Data.Signature, nil
}

// VerifyMessage verifies a message signature
func (s *WalletService) VerifyMessage(ctx context.Context, publicKey, message, signature string) (bool, error) {
	req := types.VerifyMessageRequest{
		PublicKey: publicKey,
		Message:   message,
		Signature: signature,
	}
	
	var resp types.VerifyMessageResponse
	err := s.httpClient.Post(ctx, "/wallets/verify", req, &resp)
	if err != nil {
		return false, fmt.Errorf("failed to verify message: %w", err)
	}
	
	if !resp.Success {
		return false, fmt.Errorf("failed to verify message: %s", safeString(resp.Error))
	}
	
	return resp.Data.Valid, nil
}

// SubscribeToWalletEvents subscribes to real-time wallet events
func (s *WalletService) SubscribeToWalletEvents(ctx context.Context, walletID string) (*WebSocketSubscription, error) {
	if s.wsClient == nil {
		return nil, fmt.Errorf("WebSocket client not available")
	}
	
	filters := map[string]interface{}{
		"wallet_id": walletID,
	}
	
	return s.wsClient.Subscribe(ctx, filters)
}

// SwitchNetwork switches wallet network
func (s *WalletService) SwitchNetwork(ctx context.Context, walletID string, network types.NetworkType) (*types.Wallet, error) {
	req := map[string]interface{}{
		"network": string(network),
	}
	
	var resp types.WalletResponse
	err := s.httpClient.Post(ctx, fmt.Sprintf("/wallets/%s/switch-network", walletID), req, &resp)
	if err != nil {
		return nil, fmt.Errorf("failed to switch network for wallet %s: %w", walletID, err)
	}
	
	if !resp.Success {
		return nil, fmt.Errorf("failed to switch network: %s", safeString(resp.Error))
	}
	
	if s.config.IsLoggingEnabled() {
		log.Printf("Switched wallet %s to %s", walletID, network)
	}
	
	return &resp.Data, nil
}

// GetTransactionHistory retrieves transaction history for a wallet
func (s *WalletService) GetTransactionHistory(ctx context.Context, walletID string, limit, offset int) ([]map[string]interface{}, error) {
	params := map[string]string{
		"limit":  fmt.Sprintf("%d", limit),
		"offset": fmt.Sprintf("%d", offset),
	}
	
	var resp types.TransactionsResponse
	err := s.httpClient.Get(ctx, fmt.Sprintf("/wallets/%s/transactions", walletID), &resp, params...)
	if err != nil {
		return nil, fmt.Errorf("failed to get transaction history for wallet %s: %w", walletID, err)
	}
	
	if !resp.Success {
		return nil, fmt.Errorf("failed to get transaction history: %s", safeString(resp.Error))
	}
	
	return resp.Data, nil
}
